sources/handler.go

```diff
@@ -0,0 +1,487 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package sources
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/go-logr/logr"
+	"github.com/project-koku/korekuta-operator-go/crhchttp"
+)
+
+const (
+	// SourceTypesEndpoint The endpoint for retrieving source types.
+	SourceTypesEndpoint string = "source_types"
+
+	// OpenShiftSourceType The value to query to find the source type ID.
+	OpenShiftSourceType string = "openshift"
+
+	// ApplicationTypesEndpoint The endpoint for retrieving application types.
+	ApplicationTypesEndpoint string = "application_types"
+
+	// CostManagementAppType The value to query to find the application ID
+	CostManagementAppType string = "/insights/platform/cost-management"
+
+	// SourcesEndpoint The endpoint for retrieving and creating sources.
+	SourcesEndpoint string = "sources"
+
+	// NameFilterQueryParam The keyword for filtering by name via query parameter.
+	NameFilterQueryParam string = "filter[name]="
+
+	// SourceTypeIDFilterQueryParam The keyword for filtering by source_type_id via query parameter.
+	SourceTypeIDFilterQueryParam string = "filter[source_type_id]="
+
+	// SourceRefFilterQueryParam The keyword for filtering by source_ref via query parameter.
+	SourceRefFilterQueryParam string = "filter[source_ref]="
+
+	// ApplicationsEndpoint The endpoint for associating a source with an application.
+	ApplicationsEndpoint string = "applications"
+)
+
+// GenericMeta A data structure for the meta data in a paginated response
+type GenericMeta struct {
+	Count int
+}
+
+// SourceTypeDataItem A data structure for the source type item
+type SourceTypeDataItem struct {
+	ID   string
+	Name string
+}
+
+// SourceTypeResponse A data structure for the paginated source type response
+type SourceTypeResponse struct {
+	Meta GenericMeta
+	Data []SourceTypeDataItem
+}
+
+// SourceItem A data structure for the source type item
+type SourceItem struct {
+	ID           string
+	Name         string
+	SourceTypeID string
+	SourceRef    string
+}
+
+// SourceResponse A data structure for the paginated source response
+type SourceResponse struct {
+	Meta GenericMeta
+	Data []SourceItem
+}
+
+// ApplicationTypeDataItem A data structure for the application type item
+type ApplicationTypeDataItem struct {
+	ID   string
+	Name string
+}
+
+// ApplicationTypeResponse A data structure for the paginated application type response
+type ApplicationTypeResponse struct {
+	Meta GenericMeta
+	Data []ApplicationTypeDataItem
+}
+
+// CheckSource Determine if the source should be checked for existence or creation
+func CheckSource(logger logr.Logger, costConfig *crhchttp.CostManagementConfig) bool {
```

Comment:
Since this function/check is so similar to the check cycle function (https://github.com/project-koku/korekuta-operator-go/blob/master/controllers/costmanagement_controller.go#L287) is there anyway that we could create one function to use for both situations? I realize that this one is more nuanced in that it checks for the source name but I think we could possibly work around that.

Reply:
done.